cpp/src/arrow/CMakeLists.txt

```
@@ -876,6 +876,18 @@ if(ARROW_FILESYSTEM)
     foreach(ARROW_FILESYSTEM_TARGET ${ARROW_FILESYSTEM_TARGETS})
       target_link_libraries(${ARROW_FILESYSTEM_TARGET} PRIVATE ${AWSSDK_LINK_LIBRARIES})
     endforeach()
+
+    if(ARROW_S3_MODULE)
+      if(NOT ARROW_BUILD_SHARED)
+        message(FATAL_ERROR "ARROW_S3_MODULE without shared libarrow is not supported")
```

I find it super helpful when these messages let me copy and paste the option I need to pass to my CMake line instead of leaving me to figure out the flag myself

```suggestion
        message(FATAL_ERROR "ARROW_S3_MODULE without -DARROW_BUILD_SHARED=ON is not supported")
```